aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,951 @@
+            available_ids,
+            config,
+        })
+    }
+
+    /// Bootstrap policy:
+    ///   - If this is the very first run of Janus, insert keypairs in [`HpkeKeyState::Active`]
+    ///     for each configured ciphersuite. Detect first run by whether the `global_hpke_keys`
+    ///     table is empty.
+    ///   - For each configured ciphersuite, if there is no pending or active key, insert
+    ///     a key in [`HpkeKeyState::Pending`].
+    fn bootstrap(mut self) -> Result<Self, DatastoreError> {
+        for ciphersuite in &self.config.ciphersuites {
+            if self.initially_empty {
+                info!(?ciphersuite, "bootstrapping new database");
+                self.put(*ciphersuite, HpkeKeyState::Active)?;
+            } else if !self.keypairs.values().any(|keypair| {
+                &keypair.ciphersuite() == ciphersuite
+                    && (keypair.is_active() || keypair.is_pending())
+            }) {
+                info!(?ciphersuite, "bootstrapping new configuration");
+                self.put(*ciphersuite, HpkeKeyState::Pending)?;
+            }
+        }
+        Ok(self)
+    }
+
+    /// Rotation policy:
+    ///   - If the key has been in [`HpkeKeyState::Expired`] for at least
+    ///     [`HpkeKeyRotatorConfig::expired_duration`], delete it.
+    ///   - If the key has a ciphersuite that is not known, safely phase it out. If the key is
+    ///     in [`HpkeKeyState::Pending`], delete it immediately. If it is in
+    ///     [`HpkeKeyState::Active`], move it to expired.
+    ///   - For each configured ciphersuite, move the latest pending key to active if it has been
+    ///     pending for at least [`HpkeKeyRotatorConfig::pending_duration`]. Delete other pending
+    ///     keys.
+    ///   - For each configured ciphersuite, if the active key has been active for more than
+    ///     [`HpkeKeyRotatorConfig::active_duration`] and there is no pending key, insert a pending
+    ///     key.
+    ///   - For each configured ciphersuite, expire the oldest active keys.
+    fn sweep(mut self) -> Result<Self, DatastoreError> {
+        enum Op {
+            Create(HpkeCiphersuite),
+            Update(HpkeConfigId, HpkeKeyState),
+            Delete(HpkeConfigId),
+        }
+
+        let ops: Vec<_> = self
+            .keypairs
+            .iter()
+            .filter_map(|(&id, keypair)| {
+                let ciphersuite = self.config.ciphersuites.get(&keypair.ciphersuite());
+                match keypair.state() {
+                    HpkeKeyState::Active => match ciphersuite {
+                        Some(ciphersuite) => {
+                            let keypairs: Vec<_> = self
+                                .keypairs
+                                .values()
+                                .filter(|keypair| keypair.ciphersuite() == *ciphersuite)
+                                .collect();
+
+                            let pending_key_ready = keypairs.iter().any(|keypair| {
```

Comment:
This is pretty fragile: it needs to match the logic in the `Pending` case. You got it right (I think), but later changes could easily break this, leading to us expiring all active keys (or, less seriously, advertising two active keys for a given ciphersuite at a time).

In general, the operation for each keypair is computed entirely independently of each other keypair, so "lockstep" changes (such as promoting a pending keypair to active & simultaneously demoting the old active keypair to expired) have to "agree" in the logic of their processing. [1]

I suggest an "imperative" implementation strategy like the following. First, split the stored keypairs by ciphersuite so that you have a list of keypairs per ciphersuite, and process each ciphersuite's keypairs indepdendently. For each ciphersuite's keypairs:
1) Clone the list of keypairs (or perhaps map it into `(id, state, last_state_change_at)`).
2) Apply each policy rule iteratively, implemented by actually updating the in-memory state of the keypairs from the previous step. (An order that would work would be: promote old-enough pending to active -> demote old active to expired -> delete old expired; if implemented properly, I think we'd get the correct behavior for non-configured ciphersuites without needing special steps for that case.) This way, each update would be apparent to the later steps.
3) Diff the in-memory state against the original state to determine which operations to perform against the DB.
4) Finally, perform the DB operations.

I think this implementation strategy, other than being more obviously correct, would also be easier to follow & simpler to write down. WDYT? (If you disagree, definitely push back -- you've been staring at this problem longer than I have at this point, so if you think this is a bad idea, maybe it is.)

[1] Well, they could look at the operation computed for the other keypairs, but IMO this is more confusing than the suggested imperative update method; and this would also depend on the order than the keypairs are processed in.

Reply:
It's a more-than-plausible idea, but I'm not sure if there's enough benefit to justify additional work to iterate on this again, since the changes proposed seem rather sweeping.

Is the only problem with this code the duplicated logic between Pending and Active? If so, I think it can be factored out into a small function.

> but later changes could easily break this, leading to us expiring all active keys (or, less seriously, advertising two active keys for a given ciphersuite at a time).

Unit tests ought to catch any serious regressions like this--I think they should be covering this arm pretty well. LMK if they don't, though.

(I'd like if the defensive assertion could validate one active keypair per ciphersuite, but the bootstrap case confounds that since there could be a pending key but no active key.)

> I think we'd get the correct behavior for non-configured ciphersuites without needing special steps for that case.

I don't see how, since the policy for unconfigured ciphersuites is explicitly different than configured ones, i.e. immediate phase out without replacement. If we were treating the two the same, then yes I think this solution would be good for that.

Reply:
I've adopted your strategy after all, following @divergentdave noticing a bug related to preconditions not matching. Let me know if this is closer to what you're looking for.

Note that this implementation never writes to the in-memory state, instead we derive a list of `Op`s from the current table state, and apply them to the database. I think if we _can_ do it this way, we should, since I think the logic is easier to follow without having to consider intermediate state, i.e. we can think of `sweep()` as the "operations necessary to get the table to the desired state, given the initial state".

This implementation also turns out to be less indentation-riffic.

Reply:
> Note that this implementation never writes to the in-memory state, instead we derive a list of Ops from the current table state, and apply them to the database.

OK, well I might have to back out that change if adopting property based testing suggested by @divergentdave. To have tests complete in a reasonable amount of time, we'll have to check in-memory state rather than taking obnoxious amounts of database roundtrips.